### PR TITLE
Call PATCH inventory/device/attributes only on inventory changes

### PR DESCRIPTION
**Pull request by lluiscampos**
Make the client save a checksum of the current inventory attributes in
memory, and only call the API endpoint if the newly collected inventory
attributes differ from the already sent ones. It will always send the
inventory the first time after a boot.

Changelog: Title

### CODE REVIEW

File: app/mender.go [hunk 1]
```diff
@@ -682,19 +685,39 @@ func (m *Mender) InventoryRefresh() error {
...
-		idata = make(client.InventoryData, 0, len(reqAttr))
+	if iData == nil {
+		iData = make(client.InventoryData, 0, len(reqAttr))
 	}
-	_ = idata.ReplaceAttributes(reqAttr)
+	_ = iData.ReplaceAttributes(reqAttr)
 
-	if idata == nil {
-		log.Infof("No inventory data to submit")
+	if iData == nil {
```
Comment (oleorhagen) [inline app/mender.go:693]: Nitpick:
I usually like this in a separate commit, but no hard requirement

File: app/mender_test.go [hunk 2]
```diff
@@ -756,7 +778,7 @@ echo foo=bar`),
 	// 3. pretend client is no longer authorized
 	srv.Auth.Token = []byte("footoken")
 	err = mender.InventoryRefresh()
-	assert.NotNil(t, err)
```
Comment (oleorhagen) [inline app/mender_test.go:759]: Question:
Was this wrong previously?

File: client/inventory.go [hunk 3]
```diff
@@ -41,3 +41,11 @@ func (id *InventoryData) ReplaceAttributes(attr []InventoryAttribute) error {
 	}
 	return nil
 }
+
+// ByName implements sort.Interface based on the Name field, so that an
```
Comment (oleorhagen) [inline client/inventory.go:45]:
```suggestion
// ByName implements sort.Interface based on the Name field, so that a
```

File: client/inventory.go [hunk 4]
```diff
@@ -41,3 +41,11 @@ func (id *InventoryData) ReplaceAttributes(attr []InventoryAttribute) error {
...
+// slice of InventoryAttribute can be sorted and compared.
+type ByName []InventoryAttribute
```
Comment (oleorhagen) [inline client/inventory.go:47]: 👍

File: app/mender.go [hunk 5]
```diff
@@ -682,19 +685,39 @@ func (m *Mender) InventoryRefresh() error {
...
+	sort.Sort(client.ByName(iData))
+
+	iDataBuf, err := json.Marshal(iData)
+	if err != nil {
+		return err
+	}
+	iDataSha := sha256.Sum256(iDataBuf)
+	lastInvSha, err := m.GetCurrentInventoryHash()
+	if err == nil && iDataSha == lastInvSha {
+		log.Info("Inventory data has not changed, skip submit")
```
Comment (oleorhagen) [inline app/mender.go:708]: 🤔 
Logging is always hard. I don't particularly like this message, but I'm also struggling with finding something "better":
`The inventory data has not changed since the last time it was collected. Nothing to submit. Skipping server inventory update call`. 
Or something along these lines?

File: app/mender.go [hunk 6]
```diff
@@ -682,19 +685,39 @@ func (m *Mender) InventoryRefresh() error {
...
+		return nil
+	}
+
+	log.Debugf("New inventory data, updating database with hash: %v , inventory: %v", iDataSha, iData)
```
Comment (oleorhagen) [inline app/mender.go:712]: Question:
Cant the hash be formatted with:
`%x` ? And the inventory be stringified?

**Review by oleorhagen — COMMENTED** [review]
I only had time to do a superficial lookover now. But I think it looks good! No need to panic on spending time away from the client ;)

Comments are mostly questions, I think this is fine. Feel free to cherry pick at will. Will do a closer look tomorrow morning :)

File: app/mender.go (hunk 1 above)
Comment (lluiscampos) [reply app/mender.go:693]: Generally speaking I agree. But this time I am "using" idata/iData in my commit, so in that sense I think is justified to rename it :)

File: app/mender.go (hunk 6 above)
Comment (alfrunes) [reply app/mender.go:712]: ... or even `%064x`?

File: app/mender_test.go (hunk 2 above)
Comment (lluiscampos) [reply app/mender_test.go:759]: No, is just that checking `err` is not nil covers the same as checking that `err` is an `Error` (while IMO the later is more meaningful)

File: app/mender.go (hunk 6 above)
Comment (lluiscampos) [reply app/mender.go:712]: Hehe, `%v` is the lazy way...

I'll rework it a bit (at the end of the day, the main goal was to practice golang!)

Comment (kacf) [reply app/mender.go:712]: Yes, `%v` is generally discouraged because it silences all warnings about type mismatch, even useful ones. We have plenty of examples where we have made mistakes here.

File: app/mender.go [hunk 7]
```diff
@@ -682,19 +685,39 @@ func (m *Mender) InventoryRefresh() error {
...
-	if idata == nil {
-		log.Infof("No inventory data to submit")
+	if iData == nil {
+		log.Info("No inventory data to submit")
 		return nil
 	}
 
-	err = ic.Submit(m.api.Request(m.authToken, nextServerIterator(m), reauthorize(m)), m.Config.Servers[0].ServerURL, idata)
+	// sort slice to have a reproducible hash
+	sort.Sort(client.ByName(iData))
```
Comment (alfrunes) [inline app/mender.go:699]:
```suggestion
	sort.Slice(func(i, j) bool { return strings.Compare(iData[i], iData[j]) < 0 })
```
With this you don't have to introduce the extra `ByName` type.

**Review by kacf — APPROVED** [review]
Apart from the things already mentioned, looks good to me!

File: app/mender_test.go (hunk 2 above)
Comment (alfrunes) [reply app/mender_test.go:759]: The assertion is identical, it just makes it more obvious if it fails due to an error.

**Review by alfrunes — APPROVED** [review]
Well done!
I just have a couple of semantic suggestions.

**Comment by tranchitella** [comment]
@lluiscampos @mzedel heads-up: this PR is going to change the semantic of the "updated_ts" inventory attribute, which contains the current date/time of the latest upsert. In the UI, this field is reported as "Last check-in", and some of our customers are using it as a sort-of monitoring tool. When this PR is merged, or the equivalent in the backend (https://github.com/mendersoftware/inventory/pull/211) is merged, we won't update anymore this attribute and the "Last check-in" in the UI will have a different meaning (latest inventory update).

**Comment by mzedel** [comment]
Good point @tranchitella! This will then also break the "active devices" widget or the "liveness" check for devices - I'm not sure if that's acceptable... as long as we don't have a persistent connection between device and backend or a different way of tracking the check-in, I'm quite sure this would cause trouble...

**Comment by lluiscampos** [comment]
@tranchitella @mzedel Thanks for bringing this up. Could the UI get this information from the API call to deployments `/next`?

File: app/mender.go (hunk 7 above)
Comment (lluiscampos) [reply app/mender.go:699]: Thanks, I'll go with your suggestion.

File: app/mender.go (hunk 5 above)
Comment (lluiscampos) [reply app/mender.go:708]: You suggestion was a bit wordy, I cut the middle part:
`The inventory data has not changed since the last time it was collected. Skipping server inventory update call`

File: app/mender.go (hunk 6 above)
Comment (lluiscampos) [reply app/mender.go]: Fixed now.

**Comment by lluiscampos** [comment]
I reworked now following @oleorhagen and  @alfrunes suggestion. I won't merge this until 1) I demo it and 2) we have a clear follow-up with the frontend team in a JIRA ticket.

**Comment by tranchitella** [comment]
> @tranchitella @mzedel Thanks for bringing this up. Could the UI get this information from the API call to deployments `/next`?

No, because deployments/next doesn't write anywhere the latest check-in.
By the way, writing something to mongodb each time we get an API call is exactly what we want to avoid...

**Comment by mzedel** [comment]
if we imagine a device that has a stable inventory and no new updates, removing the timestamp functionality would lead to this being seen as a potentially offline device - besides the widget, this was supposed to be shown in the device list at some point if I remember correctly...
We might at least be able to ease the writing frequency, as the resolution of the check-ins does not necessarily have to be high... so it might be possible to go the same way etcd goes e.g. writing the checkins in batches once enough checkins have accumulated or on schedules and accepting lost writes in a single node - either through redundancy (as long as you get a quorum) or as long as the update interval is small enough to not care about some evicted nodes...

File: app/mender.go (hunk 5 above)
Comment (oleorhagen) [reply app/mender.go]: Nice! A little back and forth usually does the trick with these

File: app/mender.go (hunk 6 above)
Comment (oleorhagen) [reply app/mender.go]: Should it not be `%32x` ? Or am I being schtoopid here?

File: app/mender.go [hunk 8]
```diff
@@ -682,19 +686,39 @@ func (m *Mender) InventoryRefresh() error {
...
+	}
+	iDataSha := sha256.Sum256(iDataBuf)
+	lastInvSha, err := m.GetCurrentInventoryHash()
+	if err == nil && iDataSha == lastInvSha {
+		log.Info("The inventory data has not changed since the last time it was collected. Skipping server inventory update call")
+		return nil
+	}
+
+	log.Debugf("New inventory data, updating database with hash: %064x, inventory: %s", iDataSha, iData)
+	err = m.Store.WriteAll(datastore.InventoryHashKey, iDataSha[:])
```
Comment (lluiscampos) [inline app/mender.go:714]: This should be done after `ic.Submit`, as pointed today in Sprint Review by @pasinskim

**Comment by mirzak** [comment]
> Make the client save a checksum of the current inventory attributes its local store

A couple of questions regarding "local store".

Is "local store" the db that resides in the data partition?

Also does this really need to be stored in the db (if that is the case)? I would imagine that volatile memory should be enough, and it is ok for the device to send inventory once on each reboot/restart? Using volatile memory is preferred to avoid disk writes, which could increase significantly for the Mender client if the inventory values changes.

Could using "volatile memory" also help with the "last check-in" problem?

**Comment by oleorhagen** [comment]
> Could using "volatile memory" also help with the "last check-in" problem?

This sounds to me like a very good idea.

**Comment by lluiscampos** [comment]
> > Make the client save a checksum of the current inventory attributes its local store
> 
> A couple of questions regarding "local store".
> 
> Is "local store" the db that resides in the data partition?

Yes

> Also does this really need to be stored in the db (if that is the case)? I would imagine that volatile memory should be enough, and it is ok for the device to send inventory once on each reboot/restart? Using volatile memory is preferred to avoid disk writes, which could increase significantly for the Mender client if the inventory values changes.

This is something I considered during development. The pros and cons are exactly what you pointed out. As my goal was to optimize on API calls, I decided to save on disk. But re-thinking now after your comment I think it makes sense to just save it in memory (and hence forcing an inventory PATCH after reboot).

> 
> Could using "volatile memory" also help with the "last check-in" problem?

Well, in this case "last check-in" will correspond really to "last boot time", helps a bit but still won't work as today.

Thanks for your input @mirzak, I'll rework :)

File: app/mender.go (hunk 8 above)
Comment (lluiscampos) [reply app/mender.go]: Fixed now

Comment (lluiscampos) [reply app/mender.go]: See https://github.com/mendersoftware/mender/compare/822e0fc90fa98f45e37eecb32dee62d7eaafca3b..2a77a52b5ebd787a73788d5b68f55e1663283669

**Comment by lluiscampos** [comment]
> Thanks for your input @mirzak, I'll rework :)

Fixed now (see rework [here](https://github.com/mendersoftware/mender/compare/2a77a52b5ebd787a73788d5b68f55e1663283669..daf691de4e3e0227445749b4d4ea765f1c4b491e)) and also edited the commit message a bit.

**Comment by tranchitella** [comment]
@lluiscampos after discussion with @estenberg, his feedback is to NOT merge this PR and wait for an alternative implementation of the latest check-in in the UI: too many users are relying on this (non-)feature.

**Review by tranchitella — CHANGES_REQUESTED** [review]
As long as Mender does not offer another mean to provide the "latest check-in" feature, we cannot merge this PR.

**Comment by lluiscampos** [comment]
Closing for now, see [MEN-3981](https://tracker.mender.io/browse/MEN-3981).
